apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -1794,29 +1794,66 @@ def step(self, closure=None, *, grad_scaler=None):
                     device=self.device,
                 )
 
-        # Apply optimizer step
+        # Apply optimizer step and synchronize params
         self.state['step'] += 1
-        if self.store_param_remainders:
-            self._local_step_with_param_remainders()
-        else:
-            self._local_step()
+        if self.distributed_size > 1 and self.overlap_param_sync and self.state['buckets']:
+            # Local step and non-blocking param sync
+            # Note: Overlap param sync of first buckets with optimizer
+            # step of remaining buckets.
+
+            # Get buckets containing "first" parameter
+            fragment = self.state['buckets'][-1].fragments[-1]
+            param_group_id = fragment.param_group_id
+            param_id = fragment.param_id
+            param = self.param_groups[param_group_id]['params'][param_id]
+            first_bucket_ids = sorted(
+                fragment.bucket_id
+                for fragment in self.state[param]['fragments']
+            )
+
+            # Local step and launch param sync for first buckets
+            self._local_step(first_bucket_ids)
+            self._start_bucket_param_sync(
+                self._params_buckets[bucket_id]
+                for bucket_id in first_bucket_ids
+            )
+
+            # Local step for remaining buckets
+            first_bucket_ids = set(first_bucket_ids)
```

Comment:
this seems to allow duplicated bucket ids in the call of `_lcoal_step` above but would it be the case in the use case?